core/src/locktower.rs

```diff
@@ -0,0 +1,485 @@
+use crate::bank_forks::BankForks;
+use crate::staking_utils;
+use hashbrown::{HashMap, HashSet};
+use solana_runtime::bank::Bank;
+use solana_sdk::account::Account;
+use solana_sdk::pubkey::Pubkey;
+use solana_vote_api::vote_instruction::Vote;
+use solana_vote_api::vote_state::{Lockout, VoteState, MAX_LOCKOUT_HISTORY};
+
+const VOTE_THRESHOLD_DEPTH: usize = 8;
+const VOTE_THRESHOLD_SIZE: f64 = 2f64 / 3f64;
+
+pub struct StakeLockout {
+    lockout: u64,
+    stake: u64,
+}
+
+#[derive(Default)]
+pub struct Locktower {
+    pub max_stake: u64,
+    threshold_depth: usize,
+    threshold_size: f64,
+    lockouts: VoteState,
+}
+
+impl Locktower {
+    pub fn new_from_forks(bank_forks: &BankForks) -> Locktower {
+        //TODO: which bank to start with?
+        let mut frozen_banks: Vec<_> = bank_forks.frozen_banks().values().cloned().collect();
+        frozen_banks.sort_by_key(|b| (b.parents().len(), b.slot()));
+        if let Some(bank) = frozen_banks.last() {
+            Self::new_from_bank(bank)
+        } else {
+            Self::default()
+        }
+    }
+
+    pub fn new_from_bank(bank: &Bank) -> Locktower {
+        let mut max_stake = 0;
+        let current_epoch = bank.get_epoch_and_slot_index(bank.slot()).0;
+        let mut lockouts = VoteState::default();
+        if let Some(iter) = staking_utils::node_staked_accounts_at_epoch(bank, current_epoch) {
+            for (delegate_id, stake, account) in iter {
+                max_stake += stake;
+                if *delegate_id == bank.collector_id() {
+                    let state = VoteState::deserialize(&account.data).expect("votes");
+                    if lockouts.votes.len() < state.votes.len() {
+                        //TODO: which state to init with?
```

Comment:
this question goes away with the new staking proposal?

Reply:
@rob-solana yep

Reply:
`lockouts` is supposed to eventually represent this node's previous votes?